File: ruby/Flyweight/flight_factory.rb
require_relative 'Flight'
class FlightFactory
  def initialize
    @flight_map = Hash.new
  end

  def get_flight (flight_company)
    if @flight_map.has_key?(flight_company) # > 1
      @flight_map[flight_company]
    else
      flights = Flight.new(flight_company)
      @flight_map[flight_company] = flights # > 2
    end
  end
end

File: java/Flyweight/src/flyweight/FlightFactory.java
package flyweight;

import java.util.HashMap;
import java.util.Map;

public class FlightFactory {
	static final Map<String, Flight> flightMap = new HashMap<>(); // >1
	
	Flight getFlight(String flightKey) {
		Flight flight = flightMap.get(flightKey);
		
		if(flight == null) {
			flight = new Flight();
			flight.company = flightKey;
			flightMap.put(flightKey, flight); // > 2
		}
		
		return flight;
	}
	
}
